### DISCUSSION TITLE
How can I share exported types across multiple repos?

### DISCUSSION THREAD

**Discussion by arda-** [prisma/prisma#5526] [Q&A] [answered]
I want to generate the schema once in a back-end and use all the types on the client as well.

**Comment by nikolasburk** [comment, marked as answer]
Hey @arda-, this question was also recently discussed on Slack [here](https://prisma.slack.com/archives/CKQTGR6T0/p1612419345000200?thread_ts=1612400703.496400&cid=CKQTGR6T0) and [here](https://prisma.slack.com/archives/CA491RJH0/p1612751729344400).

Here are your options:

- If you’re using a monorepo, you can directly share the backend types by importing them in the frontend.
- If your frontend and backend projects are in different repos, it might be best to create an npm package that includes the types which you can install on the frontend and that can be imported.

A 3rd option might be to write a script that copies the schema and generates the Prisma Client types and include it in your build process.

Let me know if that helps or if you have any further questions, always happy to help :)

**Reply by MarcMogdanz** [reply]
Hey @nikolasburk!

I'm in the same boat as @arda- right now and need to access the types from multiple projects. It is indeed a monorepo but due to build/deploy constraints I can't import them directly so I wanted to go with the private npm package approach. I couldn't wrap my head around on how to import and export prisma with the built types in a separate package. There are a few discussions (#6296 #6243 #10100) already about this topic but no example yet. The only example I found was a [blog article](https://isidoro-ferreiro.medium.com/share-your-prisma-client-across-projects-44d1c7aca6fd) by @isidoro98 but which I couldn't get running.

Is there any working example out there on how to build a separate package "wrapping" prisma with the correct types that can be imported like the "normal" prisma client would be?

And also, is there a way to access the slack without having an `@outrider.ai` email address?

**Reply by nikolasburk** [reply]
Hey Marc, thanks for following up!

> Is there any working example out there on how to build a separate package "wrapping" prisma with the correct types that can be imported like the "normal" prisma client would be?

I'm not aware of any such example unfortunately, but this is certainly something we should keep on our list to create eventually! If you raise this specific point in a new GitHub Discussion and explain the errors you have been running into, maybe our Dev Success Engineers would be able to help you.

> And also, is there a way to access the slack without having an `@outrider.ai` email address?

Are you referring to the [public Prisma Slack](https://slack.prisma.io)? If so, you should be able to join it with any email addresses, I have never heard of `@outrider.ai` 😅

**Reply by MarcMogdanz** [reply]
> I'm not aware of any such example unfortunately, but this is certainly something we should keep on our list to create eventually! If you raise this specific point in a new GitHub Discussion and explain the errors you have been running into, maybe our Dev Success Engineers would be able to help you.

Sounds good, will do when I find the time for it!

> Are you referring to the public Prisma Slack? If so, you should be able to join it with any email addresses, I have never heard of @outrider.ai 😅

When I try to view the Slack links you mentioned in your initial reply, I get prompted to sign in or register and when registering I get the following screen:

![Screenshot 2021-11-25 at 00 27 35](https://user-images.githubusercontent.com/20729218/143325155-90f10cbb-2d84-4c3f-8c46-f85849ad193b.png)

But using the invite bot it seems to work now, thanks!

**Reply by MarcMogdanz** [reply]
Update: After another few hours working on it, I've got it running!

I've created an example repository with a small guide for everyone who is in the same boat. It contains an example prisma setup wrapped in a package and a Next.js app that uses this package. Bonus: guide on how to manually pack the package and also a GitHub actions workflow to publish it to a npm registry. Hope this helps :)

https://github.com/MarcMogdanz/typed-prisma-package

**Reply by nikolasburk** [reply]
> When I try to view the Slack links you mentioned in your initial reply, I get prompted to sign in or register and when registering I get the following screen:

Wow, that's weird, thanks for letting me know! Will try to look into this:)

> I've created an example repository with a small guide for everyone who is in the same boat. It contains an example prisma setup wrapped in a package and a Next.js app that uses this package. Bonus: guide on how to manually pack the package and also a GitHub actions workflow to publish it to a npm registry. Hope this helps :)

Oh neat! Would you mind if I share this on Twitter? (If you have an account there, I'd of course also be happy to mention/credit you in the Tweet).

**Reply by MarcMogdanz** [reply]
> Oh neat! Would you mind if I share this on Twitter? (If you have an account there, I'd of course also be happy to mention/credit you in the Tweet).

I don't have a twitter account but feel free to post it anyways!

**Comment by mkaradeniz** [comment]
There is a great example-repo: https://github.com/aiji42/turbo-with-prisma

This uses `webpack` with `string-replace-loader` instead of a custom script: https://github.com/aiji42/turbo-with-prisma/blob/main/apps/web/next.config.js

**Comment by JoeRoddy** [comment]
If anyone is interested, I had this same use case, so I made a generator to export your Prisma client's types to a specified directory and removes the compiled JS files.

1. `npm install -D prisma-types-export`
2. Add the following to your `schema.prisma` file:
   ```prisma
   generator types {
       provider = "prisma-types-export"
       output   = "./prisma-types"
   }
   ```
3. `npx prisma generate`

Directly copying one of `/@prisma/client` or `/.prisma/client` have pitfalls, because they both actually reference each other, this combines them into one directory and updates the import statements.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
